File: chat_mobile/lib/data/models/message.dart
// lib/data/models/message.dart

class Message {
  final String id;
  final String conversationId;
  final String senderId;
  final String? recipientUserId;
  final String? senderName;
  final String? decryptedContent;
  final String encryptedContent;
  final String? nonce;
  final String? authTag;
  final String? signature;
  final String type;
  final String? status;
  final DateTime timestamp;
  final bool isRead;
  final bool isDelivered;
  final Map<String, dynamic>? metadata;

  Message({
    required this.id,
    required this.conversationId,
    required this.senderId,
    this.recipientUserId,
    this.senderName,
    this.decryptedContent,
    required this.encryptedContent,
    this.nonce,
    this.authTag,
    this.signature,
    this.type = 'TEXT',
    this.status,
    required this.timestamp,
    this.isRead = false,
    this.isDelivered = false,
    this.metadata,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id']?.toString() ?? '',
      conversationId: json['conversation_id']?.toString() ?? 
                     json['conversation']?.toString() ?? '',
      senderId: json['sender_id']?.toString() ??
               json['from_user']?.toString() ?? '',
      recipientUserId: json['recipient_user_id']?.toString(),
      senderName: json['sender_name']?.toString(),
      encryptedContent: json['encrypted_content']?.toString() ?? '',
      nonce: json['nonce']?.toString(),
      authTag: json['auth_tag']?.toString(),
      signature: json['signature']?.toString(),
      decryptedContent: null,
      type: json['type']?.toString() ?? 'TEXT',
      status: json['status']?.toString(),
      timestamp: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : json['timestamp'] != null
              ? DateTime.parse(json['timestamp'])
              : DateTime.now(),
      isRead: json['is_read'] == true,
      isDelivered: json['is_delivered'] == true,
      metadata: json['metadata'] != null 
          ? Map<String, dynamic>.from(json['metadata'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'conversation_id': conversationId,
      'sender_id': senderId,
      if (recipientUserId != null) 'recipient_user_id': recipientUserId,
      'sender_name': senderName,
      'encrypted_content': encryptedContent,
      'nonce': nonce,
      'auth_tag': authTag,
      'signature': signature,
      'type': type,
      'status': status,
      'timestamp': timestamp.toIso8601String(),
      'is_read': isRead,
      'is_delivered': isDelivered,
      if (metadata != null) 'metadata': metadata,
    };
  }

  Message copyWith({
    String? recipientUserId,
    String? decryptedContent,
    String? status,
    bool? isRead,
    bool? isDelivered,
  }) {
    return Message(
      id: id,
      conversationId: conversationId,
      senderId: senderId,
      recipientUserId: recipientUserId ?? this.recipientUserId,
      senderName: senderName,
      decryptedContent: decryptedContent ?? this.decryptedContent,
      encryptedContent: encryptedContent,
      nonce: nonce,
      authTag: authTag,
      signature: signature,
      type: type,
      status: status ?? this.status,
      timestamp: timestamp,
      isRead: isRead ?? this.isRead,
      isDelivered: isDelivered ?? this.isDelivered,
      metadata: metadata,
    );
  }
}



// // lib/data/models/message.dart

// class Message {
//   // ═══════════════════════════════════════════════════════════════
//   // IDs
//   // ═══════════════════════════════════════════════════════════════
//   final String id;  // ✅ UUID String
//   final String conversationId;  // ✅ UUID String
//   final String senderId;  // ✅ UUID String (corrigé)
  
//   // ═══════════════════════════════════════════════════════════════
//   // CONTENU
//   // ═══════════════════════════════════════════════════════════════
//   final String? senderName;
//   final String? decryptedContent;  // ✅ Contenu déchiffré (pour affichage)
  
//   // ═══════════════════════════════════════════════════════════════
//   // CHAMPS E2EE (Backend)
//   // ═══════════════════════════════════════════════════════════════
//   final String encryptedContent;  // ✅ Base64 ciphertext
//   final String? nonce;             // ✅ Base64 nonce (12 bytes)
//   final String? authTag;           // ✅ Base64 auth tag (16 bytes)
//   final String? signature;         // ✅ Base64 Ed25519 signature
  
//   // ═══════════════════════════════════════════════════════════════
//   // MÉTADONNÉES
//   // ═══════════════════════════════════════════════════════════════
//   final String type;
//   final String? status;
//   final DateTime timestamp;
//   final bool isRead;
//   final bool isDelivered;
//   final Map<String, dynamic>? metadata;

//   Message({
//     required this.id,
//     required this.conversationId,
//     required this.senderId,
//     this.senderName,
//     this.decryptedContent,
//     required this.encryptedContent,
//     this.nonce,
//     this.authTag,
//     this.signature,
//     this.type = 'TEXT',
//     this.status,
//     required this.timestamp,
//     this.isRead = false,
//     this.isDelivered = false,
//     this.metadata,
//   });

//   // ═══════════════════════════════════════════════════════════════
//   // FROM JSON (Backend)
//   // ═══════════════════════════════════════════════════════════════
//   factory Message.fromJson(Map<String, dynamic> json) {
//     return Message(
//       id: json['id']?.toString() ?? '',
//       conversationId: json['conversation_id']?.toString() ?? 
//                      json['conversation']?.toString() ?? '',
//       senderId: json['sender_id']?.toString() ??   // ✅ String maintenant
//                json['from_user']?.toString() ?? '',
//       senderName: json['sender_name']?.toString(),
      
//       // ✅ Champs E2EE séparés
//       encryptedContent: json['encrypted_content']?.toString() ?? '',
//       nonce: json['nonce']?.toString(),
//       authTag: json['auth_tag']?.toString(),
//       signature: json['signature']?.toString(),
      
//       // Pas de contenu déchiffré depuis JSON (sera fait par service)
//       decryptedContent: null,
      
//       type: json['type']?.toString() ?? 'TEXT',
//       status: json['status']?.toString(),
      
//       timestamp: json['created_at'] != null
//           ? DateTime.parse(json['created_at'])
//           : json['timestamp'] != null
//               ? DateTime.parse(json['timestamp'])
//               : DateTime.now(),
      
//       isRead: json['is_read'] == true,
//       isDelivered: json['is_delivered'] == true,
//       metadata: json['metadata'] != null 
//           ? Map<String, dynamic>.from(json['metadata'])
//           : null,
//     );
//   }

//   // ═══════════════════════════════════════════════════════════════
//   // TO JSON
//   // ═══════════════════════════════════════════════════════════════
//   Map<String, dynamic> toJson() {
//     return {
//       'id': id,
//       'conversation_id': conversationId,
//       'sender_id': senderId,
//       'sender_name': senderName,
//       'encrypted_content': encryptedContent,
//       'nonce': nonce,
//       'auth_tag': authTag,
//       'signature': signature,
//       'type': type,
//       'status': status,
//       'timestamp': timestamp.toIso8601String(),
//       'is_read': isRead,
//       'is_delivered': isDelivered,
//       if (metadata != null) 'metadata': metadata,
//     };
//   }

//   // ═══════════════════════════════════════════════════════════════
//   // COPY WITH
//   // ═══════════════════════════════════════════════════════════════
//   Message copyWith({
//     String? decryptedContent,
//     String? status,
//     bool? isRead,
//     bool? isDelivered,
//   }) {
//     return Message(
//       id: id,
//       conversationId: conversationId,
//       senderId: senderId,
//       senderName: senderName,
//       decryptedContent: decryptedContent ?? this.decryptedContent,
//       encryptedContent: encryptedContent,
//       nonce: nonce,
//       authTag: authTag,
//       signature: signature,
//       type: type,
//       status: status ?? this.status,
//       timestamp: timestamp,
//       isRead: isRead ?? this.isRead,
//       isDelivered: isDelivered ?? this.isDelivered,
//       metadata: metadata,
//     );
//   }
// }

File: projetchat/messagerie/models/message.py



# messagerie/models/message.py
# ✅ REMPLACE TON FICHIER message.py AVEC CE CODE COMPLET

from django.db import models
from django.utils import timezone
from authentification.models import User
from .conversation import Conversation
import uuid


class Message(models.Model):
    """
    Message chiffré de bout en bout (E2EE).
    Le serveur ne peut JAMAIS lire le contenu.
    
    Architecture E2EE:
    - encrypted_content: Ciphertext AES-256-GCM
    - nonce: Nombre unique pour chaque chiffrement
    - auth_tag: Tag d'authentification GCM
    - signature: Signature Ed25519 pour authenticité
    """

    class Type(models.TextChoices):
        TEXT = "TEXT", "Text"
        IMAGE = "IMAGE", "Image"
        VIDEO = "VIDEO", "Video"
        VOICE = "VOICE", "Voice"
        FILE = "FILE", "File"
        SYSTEM = "SYSTEM", "System"

    id = models.UUIDField(
        primary_key=True,
        default=uuid.uuid4,
        editable=False
    )

    conversation = models.ForeignKey(
        Conversation,
        related_name="messages",
        on_delete=models.CASCADE
    )

    from_user = models.ForeignKey(
        User,
        related_name="sent_messages",
        on_delete=models.CASCADE
    )
    recipient_user = models.ForeignKey(
        User,
        related_name="received_messages",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        help_text="User pour qui le message a été chiffré (utilisé pour déchiffrement)"
    )

    type = models.CharField(
        max_length=20,
        choices=Type.choices,
        default=Type.TEXT
    )

    # ========================================
    # ✅ CHAMPS E2EE COMPLETS (4 champs)
    # ========================================
    
    encrypted_content = models.TextField(
        help_text="Ciphertext AES-256-GCM (Base64)"
    )
    
    nonce = models.CharField(
        max_length=255,
        default='',
        help_text="Nonce AES-GCM (Base64) - 12 bytes"
    )
    
    auth_tag = models.CharField(
        max_length=255,
        default='',
        help_text="Authentication tag AES-GCM (Base64) - 16 bytes"
    )
    
    signature = models.TextField(
        default='',
        help_text="Signature Ed25519 du hash du ciphertext (Base64)"
    )

    metadata = models.JSONField(
        blank=True,
        null=True,
        help_text="Métadonnées non sensibles (file_name, size, etc.)"
    )

    reply_to = models.ForeignKey(
        "self",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="replies"
    )

    is_deleted = models.BooleanField(default=False)

    created_at = models.DateTimeField(
        auto_now_add=True,
        db_index=True
    )

    class Meta:
        db_table = 'messages'
        ordering = ["created_at"]
        indexes = [
            models.Index(fields=["conversation", "created_at"]),
            models.Index(fields=["from_user"]),
        ]
        verbose_name = 'Message'
        verbose_name_plural = 'Messages'

    def __str__(self):
        name = self.from_user.display_name or self.from_user.phone_number
        return f"{name} - {self.type} - {self.created_at.strftime('%Y-%m-%d %H:%M')}"
    


    # from django.db import models
# from django.utils import timezone
# from authentification.models import User
# from .conversation import Conversation
# import uuid


# class Message(models.Model):
#     """
#     Message chiffré de bout en bout (E2EE).
#     Le serveur ne peut JAMAIS lire le contenu.
#     """

#     class Type(models.TextChoices):
#         TEXT = "TEXT", "Text"
#         IMAGE = "IMAGE", "Image"
#         VIDEO = "VIDEO", "Video"
#         VOICE = "VOICE", "Voice"
#         FILE = "FILE", "File"
#         SYSTEM = "SYSTEM", "System"

#     class Status(models.TextChoices):
#         SENT = "SENT", "Sent"
#         DELIVERED = "DELIVERED", "Delivered"
#         READ = "READ", "Read"

#     id = models.UUIDField(
#         primary_key=True,
#         default=uuid.uuid4,
#         editable=False
#     )

#     conversation = models.ForeignKey(
#         Conversation,
#         related_name="messages",
#         on_delete=models.CASCADE
#     )

#     from_user = models.ForeignKey(
#         User,
#         related_name="sent_messages",
#         on_delete=models.CASCADE
#     )

#     type = models.CharField(
#         max_length=20,
#         choices=Type.choices,
#         default=Type.TEXT
#     )

#     encrypted_content = models.TextField()
#     """
#     Contenu chiffré (texte, URL média chiffré, etc.)
#     """

#     metadata = models.JSONField(
#         blank=True,
#         null=True
#     )
#     """
#     Exemple:
#     {
#       "file_name": "test.pdf",
#       "size": 234234,
#       "duration": 12
#     }
#     """

#     reply_to = models.ForeignKey(
#         "self",
#         null=True,
#         blank=True,
#         on_delete=models.SET_NULL,
#         related_name="replies"
#     )

#     is_deleted = models.BooleanField(default=False)

#     created_at = models.DateTimeField(
#         auto_now_add=True,
#         db_index=True
#     )

#     class Meta:
#         ordering = ["created_at"]
#         indexes = [
#             models.Index(fields=["conversation", "created_at"]),
#             models.Index(fields=["from_user"]),
#         ]

#     def __str__(self):
#         return f"{self.from_user.display_name} - {self.type}"
    
    
# class MessageKey(models.Model):
#     """
#     Clé de message chiffrée pour chaque participant.
#     """

#     id = models.UUIDField(
#         primary_key=True,
#         default=uuid.uuid4,
#         editable=False
#     )

#     message = models.ForeignKey(
#         Message,
#         related_name="keys",
#         on_delete=models.CASCADE
#     )

#     user = models.ForeignKey(
#         User,
#         on_delete=models.CASCADE
#     )

#     encrypted_key = models.TextField()

#     class Meta:
#         unique_together = ("message", "user")
#         indexes = [
#             models.Index(fields=["user"]),
#             models.Index(fields=["message"]),
#         ]

# def __str__(self):
#     name = self.from_user.display_name or self.from_user.phone_number
#     return f"{name} - {self.type}"
